micrometer/src/test/java/feign/micrometer/MicrometerCapabilityTest.java

```diff
@@ -97,5 +110,51 @@ protected Meter getMetric(String suffix, String... tags) {
         .orElse(null);
   }
 
+  @Test
+  public void shouldTestConcurrentCreateInstances()
+      throws InterruptedException, ExecutionException {
+
+    // Given
+    final Feign.Builder builder = Feign.builder().contract(new SpringMvcContract())
+        .encoder(new JacksonEncoder())
+        .decoder(new JacksonDecoder())
+        .retryer(Retryer.NEVER_RETRY)
+        .options(new Options());
+
+
+    final MultithreadTester<Feign.Builder, ApiRestTest> tester = new MultithreadTester<>(
+        builder, 1000);
+
+    // When
+    final List<Future<ApiRestTest>> threadsExecution = tester.run(feignBuilder -> {
+
+
+      final long t1 = System.currentTimeMillis();
+
+      final ApiRestTest apiRestTest = feignBuilder
+          .addCapability(new MicrometerCapability())
+          .target(ApiRestTest.class, "http://localhost:8080");
```

Comment:
I don't think this is a valid test.

Imagine instead it was changing the `decoder`... It would be impossible to know which decoder would be used.


Same for capabilities. It's impossible to know which ones are enable for each client.

On a concurrent scenario like this one, reach client should use it's own builder.

Reply:
To be a bit more clear, we can change Feign builder to prevent this error. But, we shouldn't cause the builder is not safe to be modified by multiple threads anyway.

Only way for this to be safe, would be changing the builder to Copy on Write... so each method would return a copy of the builder.

Problem on doing that, this is a major changed on feign behaviour

Reply:
@kdavisk6 what do you think?

Reply:
I agree.  Target construction is not thread safe, and not expected to be.  You cannot share a builder across threads.  That is usage is unsupported.